### PR TITLE
Give an error when a type argument is inferred as void, and when a return statement has a void expression.

### LINKED ISSUE #1830: Compiler Happily Emits Task<void>
Source:

```C#
var t = Task.Run(() => {
	return Console.Beep();
});
```

**This should [not compile!](http://tryroslyn.azurewebsites.net/#b:master/f:sr/K4Zwlgdg5gBAygTxAFwKYFsDcAoUlaIoYB0AKgBYBOqAhgCb5k0gDWIOAbjZTMjALwxSzFsQBKwCAAopASgEA+GAG9sASGrJglCDADCAewggDAG1TEAQqlQAHOTgC+snEAA=)**

The IDE also happily reports that `t` is a `Task<void>`.
Running it throws a TypeLoadException: `The generic type 'System.Func`1' was used with an invalid instantiation in assembly mscorlib`

### PR DESCRIPTION
**Pull request by gafter**
Fixes #1830

@jaredpar @AlekseyTs @agocke @VSadov

### CODE REVIEW

**Comment by gafter** [comment]
@dotnet-bot Test this please.

**Comment by gafter** [comment]
@dotnet-bot Test this please.

**Comment by gafter** [comment]
@agocke Do you know what I have to do to trigger a Jenkins build-and-test? The previous comment did not seem to do it.

**Comment by jaredpar** [comment]
@dotnet-bot test this please

**Comment by jaredpar** [comment]
@gafter that comment was correct. I think it just coincided with a Jenkins upgrade they were doing last night and got lost in the shuffle.

**Comment by jaredpar** [comment]
@gafter hmm, just realized your request was made 17 minutes ago, not 17 hours.  No idea what was going on there.  Seems to have retriggered though now.

**Comment by AlekseyTs** [comment]
Does VB have a similar issue?

File: src/Compilers/CSharp/Portable/Symbols/ConstraintsHelper.cs [hunk 1]
```diff
@@ -647,7 +647,7 @@ private static bool InterfacesAreDistinct(NamedTypeSymbol type, ConsList<Symbol>
                 return true;
             }
 
-            if (typeArgument.IsPointerType() || typeArgument.IsRestrictedType())
+            if (typeArgument.IsPointerType() || typeArgument.IsRestrictedType() || typeArgument.SpecialType == SpecialType.System_Void)
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Symbols/ConstraintsHelper.cs]: It looks like there is no unit-test for this new code path. Consider adding one.

**Comment by gafter** [comment]
@AlekseyTs No, VB already complains if you try to return a void expression, With
```
error BC30491: Expression does not produce a value.
```

File: src/Compilers/CSharp/Portable/Symbols/ConstraintsHelper.cs (hunk 1 above)
Comment (gafter) [inline src/Compilers/CSharp/Portable/Symbols/ConstraintsHelper.cs]: I don't know how to make this happen with the current inference rules and the new error that is given on the return statement. However, this is a language constraint that might trigger in error scenarios or due to future language changes, so I believe it needs to be implemented here.

**Comment by AlekseyTs** [comment]
LGTM

**Comment by gafter** [comment]
@VSadov Could you review this please?

**Comment by VSadov** [comment]
LGTM

**Pull request merged**